### ISSUE TITLE
[DB2400] ArgumentException Column 'SQL_TYPE_NAME' does not belong to table DataTypes

### ISSUE THREAD

**Issue by shessane** [nhibernate/nhibernate-core#3438]
Using NHibernate to access DB2 AS400 database. When upgrading to NHibernate 5.3.13 from 5.2.7, we face a regression in the FluentConfiguration.BuildSessionFactory() :

> FluentNHibernate.Cfg.FluentConfigurationException : An invalid or incomplete configuration was used while creating a SessionFactory. Check PotentialReasons collection, and InnerException for more detail.
> 
> 
> ---- System.ArgumentException : Column 'SQL_TYPE_NAME' does not belong to table DataTypes.
> 
>   Stack Trace: 
> FluentConfiguration.BuildSessionFactory()
>
>  ----- Inner Stack Trace -----
> DB2MetaData.GetReservedWords()
> SchemaMetadataUpdater.GetReservedWords(Dialect dialect, IConnectionHelper connectionHelper)
> SchemaMetadataUpdater.UpdateDialectKeywords(Dialect dialect, IConnectionHelper connectionHelper)
> SchemaMetadataUpdater.Update(ISessionFactoryImplementor sessionFactory)
> SessionFactoryImpl.ctor(Configuration cfg, IMapping mapping, Settings settings, EventListeners listeners)
> Configuration.BuildSessionFactory()
> FluentConfiguration.BuildSessionFactory()

The problem seems the same as [#fluent-nhibernate/504](https://github.com/nhibernate/fluent-nhibernate/issues/504)

Here's the line that create the problem : https://github.com/nhibernate/nhibernate-core/blob/4dac4ca42841ca600e2e70803585c1f46bbe900d/src/NHibernate/Dialect/Schema/DB2MetaData.cs#L39

In a debug session I can see that there's no colomn named "SQL_TYPE_NAME". I have instead a column named "TypeName".

The workaround consists of disabling IncludeDataTypesInReservedWords :

``` csharp
// ...
fluentConfiguration.Database(DB2400Configuration.Standard.ShowSql().ConnectionString(connectionString)
    .Dialect<CustomDB2As400Dialect>())
//...

public class CustomDB2As400Dialect : DB2400Dialect
{
    public override IDataBaseSchema GetDataBaseSchema(DbConnection connection)
    {
        return new CustomAs400MetaData(connection);
    }

    private class CustomAs400MetaData : DB2MetaData
    {
        public CustomAs400MetaData(DbConnection connection) : base(connection)
        {
        }

        public override bool IncludeDataTypesInReservedWords => false;
    }
}
```

I'll try to submit a fix that returns if the column "SQL_TYPE_NAME" is not found.

**Comment by hazzik** [comment]
I'm not entirely sure how this was working in 5.2.x for you, because, in 5.2.7 both `DB2Dialect` and `DB2400Dialect` were not implementing `GetDataBaseSchema` method, and dialect is (still) throwing a `NotImplementedException`:

https://github.com/nhibernate/nhibernate-core/blob/40779f9fad8d3c9e2dd2cf0acfbbb369833b6012/src/NHibernate/Dialect/Dialect.cs#L491-L494

**Comment by hazzik** [comment]
Ok, found how, which is really bad:

https://github.com/nhibernate/nhibernate-core/blob/ac14311a9725f88116040de3baf65211f12a22b4/src/NHibernate/Impl/SessionFactoryImpl.cs#L213-L228

In short, you were using incorrect configuration in the first place. This is not a regression per se.

**Comment by hazzik** [comment]
Sounds good

**Comment by fredericDelaporte** [comment]
This is fixed in 5.3.20 by #3451 (not yet released) by removing the metadata support for DB2/400. Feel free to recycle your current PR #3439 as a new feature, supporting the metadata for DB2/400. (Better wait for the fix to be merged back to master, which will have to wait for 5.3.20 than 5.4.x release, or merge it in your branch and work over it.)

**Comment by shessane** [comment]
I'll close this one and wait for master merge. Thanks
